File: mcp-job-pipeline/lib/resume_pdf.py
import os
import shutil
import subprocess

import pdfplumber

SOFFICE_BIN = shutil.which("soffice")


def convert_docx_to_pdf(docx_path: str, outdir: str) -> dict:
    if not os.path.exists(docx_path):
        raise FileNotFoundError(docx_path)
    os.makedirs(outdir, exist_ok=True)

    result = subprocess.run(
        [SOFFICE_BIN, "--headless", "--convert-to", "pdf", "--outdir", outdir, docx_path],
        capture_output=True,
        text=True,
        timeout=60,
    )
    if result.returncode != 0:
        raise RuntimeError(f"soffice conversion failed: {result.stderr or result.stdout}")

    base = os.path.splitext(os.path.basename(docx_path))[0]
    pdf_path = os.path.join(outdir, f"{base}.pdf")
    if not os.path.exists(pdf_path):
        raise RuntimeError(f"Expected output PDF not found: {pdf_path}")

    with pdfplumber.open(pdf_path) as pdf:
        page_count = len(pdf.pages)

    return {"pdf_path": pdf_path, "page_count": page_count}

File: mcp-job-pipeline-node/lib/resumePdf.js
import fs from "node:fs";
import path from "node:path";
import { execFile } from "node:child_process";
import { promisify } from "node:util";
import { PDFDocument } from "pdf-lib";

const execFileAsync = promisify(execFile);

export async function convertDocxToPdf(docxPath, outdir) {
  if (!fs.existsSync(docxPath)) {
    throw new Error(`File not found: ${docxPath}`);
  }
  fs.mkdirSync(outdir, { recursive: true });

  try {
    // execFile resolves a bare command name via PATH without a shell, same as
    // Python's shutil.which() — no need to hunt for soffice ourselves.
    await execFileAsync("soffice", ["--headless", "--convert-to", "pdf", "--outdir", outdir, docxPath], {
      timeout: 60000,
    });
  } catch (err) {
    throw new Error(`soffice conversion failed: ${err.stderr || err.message}`);
  }

  const base = path.basename(docxPath, path.extname(docxPath));
  const pdfPath = path.join(outdir, `${base}.pdf`);
  if (!fs.existsSync(pdfPath)) {
    throw new Error(`Expected output PDF not found: ${pdfPath}`);
  }

  const pdfBytes = fs.readFileSync(pdfPath);
  const pdfDoc = await PDFDocument.load(pdfBytes);
  const pageCount = pdfDoc.getPageCount();

  return { pdf_path: pdfPath, page_count: pageCount };
}
